pkg/sentry/fs/fsutil/inode_cached.go

```diff
@@ -601,6 +605,7 @@ func (rw *inodeReadWriter) ReadToBlocks(dsts safemem.BlockSeq) (uint64, error) {
 		mr := memmap.MappableRange{uint64(rw.offset), uint64(end)}
 		switch {
 		case seg.Ok():
+			downgradelock()
```

Comment:
This is incorrect, as it's possible that `seg` only covers part of `mr`. Consider pread(offset=0, length=0x2000) where offsets [0, 0x1000) are cached and offsets [0x1000, 0x2000) are not. With the change you've made, ReadToBlocks will fill [0x1000, 0x2000) into the cache with only a read-lock on `rw.c.dataMu`. (Also, if [0x1000, 0x2000) is in cache but in a different segment, this will call rw.c.dataMu.DowngradeLock() again and panic.)

For this to be correct, you'd have to check that the entirety of mr is cached before downgrading the lock. But I'm not entirely convinced that this is an *overall* win in most cases, as this check has a cost, and in the cases where you can downgrade the lock ReadToBlocks() is just a series of memcpys anyway.